### ISSUE TITLE
Shell Integration and beyond

### ISSUE THREAD

**Issue by wez** [wez/wezterm#115]
This issue is a placeholder/umbrella for some things that are not super well defined today.

 * [x] OSC 7 for CWD and tabs; implemented in https://github.com/wez/wezterm/issues/113
 * [ ] [semantic prompts](https://gitlab.freedesktop.org/Per_Bothner/specifications/blob/master/proposals/semantic-prompts.md); see it [visualized here](http://per.bothner.com/blog/2019/shell-integration-proposal/).  That spec is derived from [the shell integration feature in iTerm2](https://iterm2.com/documentation-shell-integration.html)

**Comment by ImportTaste** [comment]
How about [starship](https://github.com/starship/starship) integration? It's also coded in Rust.

Though I'm not sure if integration in this context means embedding your own version, or making them work harmoniously.

**Comment by wez** [comment]
starship and powerline and utilities in that category customize your shell to show contextual or system information (and don't really care much about the terminal really) but this issue more about making the terminal understand the semantics of what is being output, rather than just dealing with the bytes.

The linked semantic escapes featureset has escape sequences that the shell can output and allow iTerm2 to define a keyboard shortcut to jump to the start/end of a previous command's output, rather than just by scrolling/paging backwards.

DomTerm uses the additional information to render your prompt area with different styling and gutter to make it much more visually apparent where the output and input begin and end.

Those escapes open things up for some interesting/useful additional functions; for example, rather than manually dragging to select the output from the last command, if the terminal knew its boundaries, then we could eg: offer a keyboard shortcut to copy the entire output from the last command.

**Comment by ImportTaste** [comment]
My mistake, I keep getting wzsh and wezterm conflated in my head.

**Comment by dfrankland** [comment]
Has anyone tested OSC 7 CWD and tabs for macOS? I use fish shell which supports it natively, but it doesn't seem to work for me.

**Comment by wez** [comment]
@dfrankland it's working for me using https://wezfurlong.org/wezterm/shell-integration.html#osc-7-escape-sequence-to-set-the-working-directory (and some more elaborate variations of this).

I'd suggest running `RUST_LOG=trace /path/to/wezterm` and then changing the directory in your fish shell a couple of times.
The stderr should capture the parsed version of the escape sequences and we go from there.

**Comment by dfrankland** [comment]
I finally figured it out! [I found how `fish` is updating the OSC 7 CWD](https://github.com/fish-shell/fish-shell/blob/3eff7b26bf3cc2b62d2ff7ebdc49d2793e718496/share/functions/__fish_config_interactive.fish#L279) and added that to the `fish_prompt`:

```fish
function fish_prompt
    printf \e\]7\;file://%s%s\a $hostname (string escape --style=url $PWD)
    printf '%s%s%s@%s %s%s%s>' \
        (set_color $fish_color_cwd) \
        (whoami) \
        (set_color normal) \
        (hostname | cut -d . -f 1) \
        (set_color $fish_color_cwd) \
        (prompt_pwd) \
        (set_color normal)
end
```

**Comment by wez** [comment]
@dfrankland thanks for that; I've opened a PR to have fish detect wezterm so that this should be more automatic in a future release of fish!

**Comment by wez** [comment]
@dfrankland BTW, we've since grown https://github.com/wez/wezterm/blob/master/assets/shell-integration/wezterm.sh with deeper integration for bash and zsh.  In addition to OSC 7 there are some OSC 133 sequences that can be used to semantically markup the prompt "chrome", the input text you write into the shell prompt and the output text from a command so that WezTerm knows what kind of output is where in the scrollback.   WezTerm can use those annotations to "page up" and jump to the start of each command output, or to eg: change triple-click to select the entire command output.

I don't know if any of that is of interest to you, or if you fancy taking a crack at applying that to your fish config.  There's a little bit more documentation in https://wezfurlong.org/wezterm/shell-integration.html that links to relevant config options to take advantage of it, but it could do with a bit more exposition.

**Comment by Strykar** [comment]
So how do I integrate https://github.com/wez/wezterm/blob/main/assets/shell-integration/wezterm.sh with wezterm?

**Comment by wez** [comment]
> So how do I integrate https://github.com/wez/wezterm/blob/main/assets/shell-integration/wezterm.sh with wezterm?

The Fedora and Debian packages install that file to `/etc/profile.d` which should cause bash/zsh to source it on startup.

If you're not using those packages/systems, you can manually source that file from your shell startup files (eg: .bashrc)

**Comment by Strykar** [comment]
> 
> If you're not using those packages/systems, you can manually source that file from your shell startup files (eg: .bashrc)

I'm on Arch linux, I only wish to source it for wezterm, not any other terminals I use.

**Comment by wez** [comment]
> > If you're not using those packages/systems, you can manually source that file from your shell startup files (eg: .bashrc)
> 
> I'm on Arch linux, I only wish to source it for wezterm, not any other terminals I use.

The packaging for arch doesn't distribute those files today. https://matrix.to/#/!PirwUBcuIlTXwNveYz:matrix.org/$GKCp1cALTteojKA0Uz6ycCgyJKwX4MnEUJdJA9hXp7c?via=matrix.org&via=butterfly.zone&via=kwatra.me is a discussion and a patch set for adding it to the AURs.

You can however download that shell script and source it for yourself.

**Comment by yashpalgoyal1304** [comment]
_will hide this comment myself_

> semantic prompts; see it [visualized here](http://per.bothner.com/blog/2019/shell-integration-proposal/).

> The main idea is that the shell sends extra escape sequences to mark the various parts of a command: prompt, input, and output.

This above idea of "semantic prompts" is awesome, i've been thinking about something like this for quite a while, but didnt know about it. nice to know.

**Comment by rodolf0** [comment]
Hi, I like the idea of semantic prompts :D wondering if there's any prototype work being done in that realm.

It'd be really nice to be able to shade background of different commands slightly different. Or being about to jump-scroll across large command outputs. I currently hack this by making my PS1 output a non visible unlikely char and back search for that, but first class support would be really nice.

**Comment by wez** [comment]
@rodolf0 if you source that wezterm.sh file then you will have working semantic prompts as mentioned at the top of https://wezfurlong.org/wezterm/shell-integration.html#shell-integration

**Comment by rodolf0** [comment]
Nice! I'll try that out. Does each semantic section have access to any
variables bound to their value at the time of execution? (I mean probably
not, I'm just thinking about how I'd like to use it).

For example if the PWD at the time of execution is captured it enable
interesting things for hyperlinks, like completing relative paths.

Other variables that would be useful are timestamp of start/end,
pipeline-status exit codes, user, maybe some other en vars

**Comment by kaddkaka** [comment]
What part of [semantic prompts](https://gitlab.freedesktop.org/Per_Bothner/specifications/blob/master/proposals/semantic-prompts.md) is missing for this to be closed?

**Comment by wez** [comment]
Are you volunteering?  There's some stuff around capturing and recording process status and some other metadata.  Maybe it's not worth capturing but it needs some thought that I don't have time for right now.

**Comment by alexjuda** [comment]
> > So how do I integrate https://github.com/wez/wezterm/blob/main/assets/shell-integration/wezterm.sh with wezterm?
> 
> The Fedora and Debian packages install that file to `/etc/profile.d` which should cause bash/zsh to source it on startup.
> 
> If you're not using those packages/systems, you can manually source that file from your shell startup files (eg: .bashrc)

Thanks so much for this tip! I couldn't make wezterm remember my working dir when opening new tabs. My set up: wezterm installed by flatpak, Fedora 37, bash. To make matters worse, the [docs section about escape sequences](https://wezfurlong.org/wezterm/shell-integration.html#osc-7-escape-sequence-to-set-the-working-directory) was a red herring: I've already had `/etc/profile.d/vte.sh` in place, but it exited quickly because `$VTE_VERSION` wasn't set.

Manually copying and pasting [wezterm.sh](https://raw.githubusercontent.com/wez/wezterm/main/assets/shell-integration/wezterm.sh) under `/etc/profile.d/` fixed it!

**Comment by ValentinLeTallec** [comment]
Any thought on adding [DomTerm's shell-integration.fish](https://github.com/PerBothner/DomTerm/blob/master/tools/shell-integration.fish) to https://github.com/wez/wezterm/blob/main/assets/shell-integration ?

I wanted to make [SelectTextAtMouseCursor 'SemanticZone'](https://wezfurlong.org/wezterm/config/lua/keyassignment/SelectTextAtMouseCursor.html) work with fish and the documentation for shell-integration mentioned [DomTerm](https://github.com/PerBothner/DomTerm). Looking around in DomTerm's repo I found the file in question and sourcing it does enable SemanticZone to work properly. I think it could be cool to have it work out of the box for other users.

(The license is also pretty loose https://github.com/PerBothner/DomTerm/blob/master/COPYING)

**Comment by wez** [comment]
I don't use fish, haven't looked at it, and am unlikely to spend the time on fish myself. I am open to reviewing a PR that adds fish shell integration with equivalent functionality to that of the zsh/bash integration.

**Comment by eugenesvk** [comment]
FYI I've added these integrations to the python-based xonsh shell, https://github.com/jnoortheen/xontrib-term-integrations/pull/6 (this plugin is terminal-agnostic, so you can use the same shell config in iTerm and WezTerm), with all the 3 document escape code types for WezTerm
Works with a few bugs (filed with the weterm/xonsh's projects)

**Comment by eugenesvk** [comment]
As a follow-up from this discussion https://github.com/wez/wezterm/discussions/3130, would you please keep track of the following feature in the semantic zone implementation:

- Right/Multiline prompts, specifically, the ability to ignore line continuation marks '...' while being able to select all the continuation inputs together as a single zone (but only the actual input part till ␤, not the background space to the right)

**Comment by mkpaz** [comment]
FYI. wezterm.sh breaks Konsole + starship integration. I spent an hour to find out where this thin vertical line comes from, cause I had no idea wezterm installs anything to `/etc/profile.d/`. Thanks for the envs to bypass, but I'd be good either to apply such integration to wezterm only or make it manual to make users aware of its existence and potential issues like this one. 

![image](https://github.com/wez/wezterm/assets/25181998/e6bb8e97-77f9-4afd-80d7-2cbbe7d61610)

**Comment by wez** [comment]
@mkpaz I suspect that is Konsole marking up the output regions based on the semantic zones. I don't think it is broken per-se, but perhaps just not what you expected? In other words, I think Konsole also supports the same sequences, but the appearance of their effects is surprising to you.

You can disable it if you wish; see the comments in the shell integration itself:

https://github.com/wez/wezterm/blob/b54b0ce40428a8f910cd7c6fd94fd58e2f0ff88b/assets/shell-integration/wezterm.sh#L6-L15

**Comment by ninjalj** [comment]
In Konsole, this is under profile configuration, on General → Semantic Integration → Alternating bars

**Comment by arthurgeek** [comment]
@curable-online have you got it working? I sourced the script above and I can see OSC 133 sequences are set, but can't get wezterm to select according to a semantic zone.

```console
~ $ fish_prompt | cat -v
^[[38;2;249;226;175m~^[(B^[[m ^[(B^[[m$ 
~ $ source semantic-prompt.fish
~ $ fish_prompt | cat -v
^[]133;P;k=i^G^[[38;2;249;226;175m~^[(B^[[m ^[(B^[[m$ ^[]133;B^G
~ $
```

wezterm.lua:

```lua
config.mouse_bindings = {
    {
        event = { Down = { streak = 3, button = "Left" } },
        action = wezterm.action.SelectTextAtMouseCursor("SemanticZone"),
        mods = "NONE",
    },
}
```

and it doesn't work :/

**Comment by arthurgeek** [comment]
@curable-online thanks! It's just like I have it.

> You do not need to source anything, just put it in /etc/fish/config.d/ with any filename with .fish extension.

Yeah, I know, I sourced it in my output because I wanted to show that it didn't had any sequences prior to it.

> I can scroll through semantic zones properly, and also select the whole semantic zone between prompts.

Can you let me know how do you scroll through those zones and select them? Perhaps I'm doing something wrong 😕. With that mouse binding, my expectation is that a 3 left mouse click (I'm actually using a trackpad, but if I change from `SemanticZone` to `Line`, it works, so I don't think this is the issue also) in any word/letter in the output of a command would copy the full output. is that correct? I'm testing with a simple `curl -v http://ifconfig.me` command and then, then a 3 left mouse clicks on any part of the output. I'm using wezterm nightly (20240111-073635-6c36a4dd).

I also loaded zsh and [wezterms's shell integration](https://github.com/wez/wezterm/blob/main/assets/shell-integration/wezterm.sh) to see if it was an issue with fish, but it also doesn't work.

Finally, in fish I use starship, and when testing in ZSH, my ZSH config was entirely blank, but I confirmed that OSC 133 sequences were being set after sourcing wezterm's shell integration.

**Comment by mgpinf** [comment]
@arthurgeek I guess the issue for you is because of where the starship init code is being called. Based on where it is called, the shell integration script may be sourced before or after it.

In the shell integration script, the `fish_prompt` function is defined.
In the starship init script, the `fish_prompt` function is defined as well.

Source the starship init code first, then the shell integration script.

By sourcing the starship init code first, the `fish_prompt` function is now the one as defined by starship init code.

Upon sourcing the shell integration script, it makes a copy of the `fish_prompt` function to `_fishprompt_saved_prompt` function, and then redefines the `fish_prompt` function and internally calls `_fishprompt_saved_prompt` and adds shell integration escape sequences around it.

So, in case you define both starship and shell integration scripts in `conf.d` directory, ensure that the starship script name comes lexicographically before the shell integration script.

I had been facing the same issue, took quite a while to figure this out.

Hope this really helps the fish shell users using wezterm :)

**Comment by ArchWand** [comment]
Hi, are there any plans to add a "jump cursor on mouse click" functionality like iTerm2 or kitty? Clicking on the input should send a bunch of left (or right) arrow key presses to the shell to move the cursor to where the mouse clicked. Is this possible with the current version of wezterm?

**Comment by gusutabopb** [comment]
The fish script mentioned above (i.e. https://github.com/PerBothner/DomTerm/blob/master/tools/shell-integration.fish) sort of works, but it breaks non-zero `$status` on my fish prompt (more specifically, the `__fish_print_pipestatus` built-in function).

On errors, my prompt normally shows as:

```fish
⋊> ~/.conf/fish/conf.d> ls e
ls: e: No such file or directory
⋊> ~/.conf/fish/conf.d [1]>
```

But when using the shell integration script above, the non-zero status codes go missing (no red `[1]`):

```fish
⋊> ~/.conf/fish/conf.d> ls e
ls: e: No such file or directory
⋊> ~/.conf/fish/conf.d>
```


I spent some time trying to debug it, but in the end I gave up and just switched to the script [provided by iTerm2](https://iterm2.com/documentation-shell-integration.html) and placed it in `~/.config/fish/conf.d/`.

**Comment by arthurgeek** [comment]
@curable-online oh, thanks for investigating this (and sorry for the long delay here, somehow missed this reply). is this still working with the latest nightly? I'm using `wezterm 20241007-103714-ed430415` here and still can't make it work. in my previous tests I was already sourcing the semantic prompt after starfish, and OSC 133 sequences were being set, but can't make it work in `wezterm`. I also tried ZSH (without starship) and [wezterms's shell integration](https://github.com/wez/wezterm/blob/main/assets/shell-integration/wezterm.sh), to rule out it being fish/starship but it still does not work :/. I use Mac OS, btw.

**Comment by arthurgeek** [comment]
ok, finally figured this out: when wezterm [multiplexing via unix domains](https://wezfurlong.org/wezterm/multiplexing.html#unix-domains) is setup, `SelectTextAtMouseCursor("SemanticZone")` does not work. this is also the case for `bash` or `zsh` when using [wezterms's shell integration](https://github.com/wez/wezterm/blob/main/assets/shell-integration/wezterm.sh).

I could get it working only after removing:

``` 
config.unix_domains = {
	{
		name = "unix",
	},
}

config.default_gui_startup_args = { "connect", "unix" }
```

from my wezterm config.

@wez is that expected?

**Comment by arthurgeek** [comment]
a bit more info:

when not connected to a unix domain, running the following code in a debug overlay repl:

```
> window:active_tab():panes()[1]:get_semantic_zones()
[
    {
        "end_x": 3,
        "end_y": 1,
        "semantic_type": "Prompt",
        "start_x": 0,
        "start_y": 1,
    },
    {
        "end_x": 128,
        "end_y": 1,
        "semantic_type": "Output",
        "start_x": 4,
        "start_y": 1,
    },
]
>
```

while when connected to a unix domain:

```
> window:active_tab():panes()[1]:get_semantic_zones()
[]
>
```

**Comment by Foleychris** [comment]
Not sure if it should be a separate issue, but I’m also seeing the same behaviour as @arthurgeek when attempting to use `SelectTextAtMouseCursor("SemanticZone")` when connected to an ssh domain as well.

**Comment by valpackett** [comment]
PSA: Fish 4.0 emits OSC 133 sequences by itself, no scripts needed!

It [currently includes](https://github.com/fish-shell/fish-shell/blob/b431873ce55a9bc0e369c2c14f1098a3a47fa46b/src/terminal.rs#L392-L406): `click_events=1` for the prompt, `cmdline_url` for the command, and exit status. Missing continuation/right prompt support, I guess, but I'm not sure if anything interesting could be done with those. Quadruple click does select just the left prompt + the command for me, with both single (avoiding rprompt) and multi-line entries.

**Comment by PrayagS** [comment]
Can confirm the behavior that @arthurgeek described on nightly (`wezterm 20250514-071422-909573fa`).
